### DISCUSSION TITLE
Using a controlled field array remove function fires errors

### DISCUSSION THREAD

**Discussion by gilneto8** [react-hook-form/react-hook-form#7208] [Q&A] [answered]
Hey everyone.

I'll try once again to have two approaches to this, a TLDR and a not-so-TLDR.

**TLDR**: I have a controlled field array (using `useFieldArray` and `watch` [as per documentation](https://react-hook-form.com/api/usefieldarray)) and a simple structure to render these elements. Whenever I click on the button that should fire the `remove` function for the current index, it shoots out a `uncaught-error TypeError: data is not iterable` to the console.

-------------------------------

**Not TLDR**:

In the context of [this discussion](https://github.com/react-hook-form/react-hook-form/discussions/6797), now I had to add support for arrays on a controlled form environment. The 'controlled' part of it is done and tested, this problem here is a bit different and, apparently, simpler.

As stated above, I'm using the `useFieldArray` hook in junction with the `watch` function from the form to create a controlled field array:

```typescript
  const { fields, append, remove } = useFieldArray<T>({
    keyName: 'id',
    name: name as ArrayPath<T>,
    control,
  });

  const watchField = watch(name);
  const controlledArray = fields.map((field, index) => ({
    ...field,
    ...watchField[index],
  }));
```

From here, the idea would be to render the `fields` and, on each of them, have a button that invokes `remove`, while at the button have another button to invoke `append`. Simple enough. This is the code for it:

```typescript
  {/*...*/}
      <ListGroup>
        {controlledArray.map((item, index) => (
          <ListGroup.Item key={item['id']}>
            <InputGroup>
              <Form.Control
                disabled={disabled}
                defaultValue={item['name']}
                {...register(`${name}.${index}.name` as const)}
              />
              <InputGroup.Append>
                <Button
                  disabled={disabled}
                  variant={'danger'}
                  type={'button'}
                  onClick={() => remove(index)}
                >
                  X
                </Button>
              </InputGroup.Append>
            </InputGroup>
          </ListGroup.Item>
        ))}
      </ListGroup>
      {showError(name)}
      <Button
        disabled={disabled}
        className={'mt-0'}
        type={'button'}
        onClick={() => append({ name: undefined as unknown as string })}
      >
        +
      </Button>
  {/*...*/}
```

I've done iterations like these multiple times, but not with RHF and not with controlled field arrays - obviously. With the current code, I can click the `+` button any times I want without a worry, it works. But as soon as I click the `X` button to remove any of the entries, this error pops up:

![image](https://user-images.githubusercontent.com/15315355/144579908-0909df60-d4ae-44e0-b88a-0d6906fe43d7.png)

It seems obvious where the error is, but I can't seem to fix it. On top of this, there's also this case, where I add a couple of entries like the example below:

![image](https://user-images.githubusercontent.com/15315355/144580083-1bc22166-ae50-41d7-9f88-c99bc17ca99e.png)

If I now delete, for example, entry `2`, after the `data is not iterable` error pops up, this is the result:

![image](https://user-images.githubusercontent.com/15315355/144580193-34e8b98d-f5e3-47cc-a4ac-535020aaa048.png)

So it's not even recognizing the indexes, apparently.
I'm not sure what's causing this, so any idea is welcome right now. I might be dealing with two different problems, but I'm not entirely sure. 

Thanks!

**Comment by gilneto8** [comment, marked as answer]
So I found the problem, just haven't looked into why it happened yet.
Apparently, there was a problem with version '7.12.2' that was, in the meantime, fixed. Updated the package to '7.21.0' and it's working as expected.
